File: ios/Sources/WattRouter/ToolBox.swift
// ToolBox.swift — the tools a turn may call, and getting to one of them.
//
// History
//   2026-08-07  A. Sigdel  Created.
//
// Contents
//   ToolBox  Dispatch a call by name.
//
// Small, and almost all of it is about failing usefully. `run` returns a result
// for every outcome but one: an unknown name, arguments that will not decode and
// a tool that threw are all things the model can act on once it is told, and a
// turn that ended instead has discarded a next move it already had.
//
// The exception is cancellation, rethrown deliberately. Reported as a result,
// "the tool was cancelled" is something a model answers by trying again.

import Foundation

/// The tools a turn may call.
public struct ToolBox: Sendable {
    /// In the order given. Kept alongside the map because the order is what the
    /// model is shown and a dictionary has none.
    public let tools: [any Tool]
    private let byName: [String: any Tool]

    /// - Parameter tools: what the turn may call. A repeated name keeps the first
    ///   and drops the rest, so what is dispatched and what is advertised cannot
    ///   disagree — which they would if the map dropped one and the list kept both.
    public init(_ tools: [any Tool]) {
        var kept: [any Tool] = []
        var byName: [String: any Tool] = [:]
        for tool in tools where byName[tool.name] == nil {
            byName[tool.name] = tool
            kept.append(tool)
        }
        self.tools = kept
        self.byName = byName
    }

    /// The tool of that name, if there is one.
    public subscript(name: String) -> (any Tool)? { byName[name] }

    /// Run what the model asked for.
    ///
    /// - Returns: what to tell the model, always — including that the request
    ///   made no sense.
    /// - Throws: `CancellationError`, and nothing else.
    public func run(_ call: ToolCall) async throws -> ToolResult {
        guard let tool = byName[call.name] else {
            // The alternatives, not just the mistake: a model that guessed a
            // plausible name is one list away from the right one.
            let known = tools.map(\.name).sorted().joined(separator: ", ")
            return ToolResult(
                id: call.id,
                content: "there is no tool called \(call.name). Available: \(known)",
                isError: true)
        }

        do {
            return ToolResult(id: call.id, content: try await tool.run(
                arguments: Data(call.arguments.utf8)))
        } catch is CancellationError {
            throw CancellationError()
        } catch let error as DecodingError {
            // The common failure by a wide margin, and worth spelling out: the
            // model wrote these arguments and can write them again.
            return ToolResult(
                id: call.id,
                content: """
                    \(call.name) could not read its arguments: \(Self.describe(error)).
                    They must match this schema: \(tool.schema)
                    """,
                isError: true)
        } catch {
            return ToolResult(
                id: call.id, content: "\(call.name) failed: \(error.localizedDescription)",
                isError: true)
        }
    }

    /// A decoding failure in a sentence the model can act on.
    ///
    /// `localizedDescription` on a `DecodingError` says only that the data could
    /// not be read, naming neither the key nor the problem. The associated values
    /// carry both.
    private static func describe(_ error: DecodingError) -> String {
        switch error {
        case .keyNotFound(let key, _):
            "\(key.stringValue) is missing"
        case .typeMismatch(let type, let context):
            "\(Self.path(context)) should be \(type)"
        case .valueNotFound(let type, let context):
            "\(Self.path(context)) is null, and a \(type) was needed"
        case .dataCorrupted(let context):
            context.codingPath.isEmpty
                ? "it is not valid JSON" : "\(Self.path(context)) is not valid"
        @unknown default:
            "it did not match the schema"
        }
    }

    /// Where in the arguments a fault was, as the model wrote them.
    private static func path(_ context: DecodingError.Context) -> String {
        context.codingPath.isEmpty
            ? "the arguments" : context.codingPath.map(\.stringValue).joined(separator: ".")
    }
}

File: android/core/src/main/kotlin/com/getlora/wattrouter/ToolBox.kt
// ToolBox.kt: the tools a turn has, and dispatching to one.
//
// History
//   2026-08-08  A. Sigdel  Created.
//
// Every failure becomes something the model can read. An unknown name lists
// the alternatives; anything thrown becomes a sentence naming the tool. A model
// told only "error" makes the same call again.
//
// definitions() is what the model is shown, and #319 is the standing warning:
// on iOS the equivalent is reached only from tests, so every registered tool is
// invisible. #341 made `tools` a parameter of Inference.complete so this side
// cannot repeat that; this is the other end of the wire.

package com.getlora.wattrouter

import kotlinx.coroutines.CancellationException
import kotlinx.serialization.json.Json
import kotlinx.serialization.json.JsonObject
import kotlinx.serialization.json.add
import kotlinx.serialization.json.buildJsonArray
import kotlinx.serialization.json.buildJsonObject
import kotlinx.serialization.json.put

/** The tools a turn has. */
class ToolBox(tools: List<Tool>) {
    /**
     * In the order given, which is the order the model is shown. A duplicate
     * name keeps the first: two tools answering to one name is a wiring
     * mistake, and picking the later one would depend on assembly order.
     */
    val tools: List<Tool> = tools.distinctBy { it.name }

    private val byName = this.tools.associateBy { it.name }

    operator fun get(name: String): Tool? = byName[name]

    /**
     * Run what the model asked for.
     *
     * # Rely
     * Called one at a time from the turn loop, in the order the model asked.
     *
     * Never throws except [CancellationException]: every other failure is a
     * result the model can read. A dead turn is worse than a sentence.
     */
    suspend fun run(call: ToolCall): ToolResult {
        val tool = byName[call.name] ?: return ToolResult(
            id = call.id,
            // The alternatives, not just the mistake. A model that misremembers
            // a name usually recognises the right one when it sees it.
            content = "there is no tool called ${call.name}. " +
                "Available: ${byName.keys.sorted().joinToString(", ")}",
            isError = true,
        )

        return try {
            // answer rather than run, which is where a tool that captures
            // something puts it. Every tool but that one inherits the default,
            // so this line is the whole of what the change costs here.
            val answered = tool.answer(call.arguments)
            ToolResult(call.id, answered.text, answered.images)
        } catch (e: CancellationException) {
            // The one thing that propagates; see Tool.kt.
            throw e
        } catch (e: Exception) {
            ToolResult(
                id = call.id,
                content = "${call.name} failed: ${e.message ?: e::class.java.simpleName}. " +
                    "Its arguments must match this schema: ${tool.schema}",
                isError = true,
            )
        }
    }

    /**
     * What the model is shown, as the provider's `tools` array.
     *
     * @throws IllegalArgumentException if a tool's schema is not a JSON
     *   object. Loud, rather than dropping it: a tool quietly missing from the
     *   list is one the model never calls and nobody notices.
     */
    fun definitions(): String = buildJsonArray {
        tools.forEach { tool ->
            add(
                buildJsonObject {
                    put("type", "function")
                    put(
                        "function",
                        buildJsonObject {
                            put("name", tool.name)
                            put("description", tool.purpose)
                            put("parameters", tool.schema.asSchema(tool.name))
                        },
                    )
                },
            )
        }
    }.toString()

    private companion object {
        val json = Json { ignoreUnknownKeys = true }
        fun String.asSchema(tool: String): JsonObject =
            runCatching { json.parseToJsonElement(this) as JsonObject }.getOrElse {
                throw IllegalArgumentException("$tool's schema is not a JSON object: $this")
            }
    }
}
